### PR TITLE
Defer IP detection to the user

### PR DESCRIPTION
**Pull request by RealOrangeOne**
Fixes https://github.com/torchbox/django-basic-auth-ip-whitelist/issues/18

The correct way to do this varies so much, that it's not possible to implement 1 universal implementation

### CODE REVIEW

File: baipw/middleware.py [hunk 1]
```diff
@@ -61,11 +61,10 @@ def _basic_auth_response(self, request):
             return self.get_response_class()(request=request)
 
     def _get_client_ip(self, request):
-        function_path = getattr(settings, "BASIC_AUTH_GET_CLIENT_IP_FUNCTION", None)
-        func = get_client_ip
-        if function_path is not None:
-            func = import_string(function_path)
-        return func(request)
+        function_path = getattr(
+            settings, "BASIC_AUTH_GET_CLIENT_IP_FUNCTION", "baipw.utils.get_client_ip"
```
Comment (RealOrangeOne) [inline baipw/middleware.py:65]: Note: It might look weird to hard code a function reference here to our own module. I'm reworking how settings are managed here to help make this clearer, but that'll be a separate PR.

File: baipw/middleware.py [hunk 2]
```diff
@@ -61,11 +61,10 @@ def _basic_auth_response(self, request):
...
+        )
+        return import_string(function_path)(request)
```
Comment (tm-kn) [inline baipw/middleware.py:67]: **Suggestion:** I think we could make it so if there's no function_path set up, we just skip the check completely.

File: README.rst [hunk 3]
```diff
@@ -84,12 +86,27 @@ Example settings
...
-Getting IP
-~~~~~~~~~~
+Getting IP Address
+~~~~~~~~~~~~~~~~~~
+
+By default, ``BasicAuthIPWhitelistMiddleware`` uses ``request.META["REMOTE_ADDR"]``
+as the client's IP, which corresponds to the IP address connecting to Django.
+If you have a reverse proxy (eg ``nginx`` in front), this will result in the IP address of
+``nginx``, not the client.
+
```
Comment (tm-kn) [inline README.rst:96]: **Suggestion:** My opinion is: let's not introduce the default that is just a bad default. Let's completely disable the IP check unless the function is configured.

File: README.rst [hunk 4]
```diff
@@ -84,12 +86,27 @@ Example settings
...
+Correctly determining the IP address can vary between deployments. Guessing incorrectly can
+result in security issues. Instead, this library requires you configure this yourselves.
```
Comment (tm-kn) [inline README.rst:98]: **Thoughts:** We should completely remove the `BASIC_AUTH_WHITELISTED_IP_NETWORKS` setting, and the "get client ip" function. We should probably call it just a `BASIC_AUTH_EXTRA_CHECK_FUNCTION` or something like that, and make it return True if the request passes the check. That way this package stops doing the IP checks explicitly altogether.

We should add a recipe in the README for adding an IP check though.

File: baipw/middleware.py (hunk 1 above)
Comment (tm-kn) [reply baipw/middleware.py:65]: I think this should return `None` by default and we should not have any default implementation.

**Review by tm-kn — CHANGES_REQUESTED** [review]
Hey, I've added some comments. Let me know what you think.

File: baipw/tests/test_utils.py [hunk 5]
```diff
@@ -43,39 +43,3 @@ def test_authorise_with_invalid_header(self):
...
-
-class TestGetClientIP(TestCase):
-    def setUp(self):
-        self.request = RequestFactory().get("/")
-
-    def test_get_client_ip_from_remote_addr(self):
-        self.request.META["REMOTE_ADDR"] = "192.168.0.17"
-        self.assertNotIn("HTTP_X_FORWARDED_FOR", self.request.META)
-        self.assertIn("REMOTE_ADDR", self.request.META)
-        self.assertEqual(get_client_ip(self.request), "192.168.0.17")
```
Comment (tm-kn) [inline baipw/tests/test_utils.py:56]: **Question:** Should we not keep the test for `REMOTE_ADDR`?

File: baipw/utils.py [hunk 6]
```diff
@@ -7,26 +7,12 @@
...
-        return
+    """
+    Get the client's IP address
 
-    # If there is a list of IPs provided, use the last one (should be
-    # the most recent one). This may not work on Google Cloud.
-    return final_ip.split(",")[-1].strip()
+    Note: This is the address connecting to Django, which is likely incorrect.
+    """
+    return request.META.get("REMOTE_ADDR")
```
Comment (tm-kn) [inline baipw/utils.py:15]: Given this can return `None` technically, should we make sure to add a None check in the `_get_client_ip` method?

**Review by tm-kn — CHANGES_REQUESTED** [review]

File: baipw/tests/test_utils.py (hunk 5 above)
Comment (RealOrangeOne) [reply baipw/tests/test_utils.py:56]: There are existing tests which cover that `REMOTE_ADDR` is used.

File: baipw/utils.py (hunk 6 above)
Comment (RealOrangeOne) [reply baipw/utils.py:15]: I'd always assumed `REMOTE_ADDR` would always exist, but according to Django it's not a 100% guarantee, so I agree I'll add some tests!.

**Review by tm-kn — APPROVED** [review]
Thanks @RealOrangeOne! Looking good. 🙌
